pkg/admin/data/data_manager.go

```diff
@@ -0,0 +1,151 @@
+package data
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/openshift/microshift/pkg/config"
+	"github.com/openshift/microshift/pkg/util"
+	"k8s.io/klog/v2"
+)
+
+var (
+	cpArgs = []string{
+		"--verbose",
+		"--recursive",
+		"--preserve",
+		"--reflink=auto",
+	}
+)
+
+func NewManager(storage StoragePath) (*manager, error) {
+	if storage == "" {
+		return nil, &EmptyArgErr{argName: "storage"}
+	}
+	return &manager{storage: storage}, nil
+}
+
+var _ Manager = (*manager)(nil)
+
+type manager struct {
+	storage StoragePath
+}
+
+func (dm *manager) GetBackupPath(name BackupName) string {
+	return filepath.Join(string(dm.storage), string(name))
+}
+
+func (dm *manager) BackupExists(name BackupName) (bool, error) {
+	return pathExists(dm.GetBackupPath(name))
+}
+
+func (dm *manager) Backup(name BackupName) error {
+	klog.InfoS("Backing up the data",
+		"storage", dm.storage, "name", name, "data", config.DataDir)
+
+	if name == "" {
+		return &EmptyArgErr{"name"}
+	}
+
+	if found, err := pathExists(string(dm.storage)); err != nil {
+		return err
+	} else if !found {
+		if makeDirErr := util.MakeDir(string(dm.storage)); makeDirErr != nil {
+			return fmt.Errorf("making %s directory failed: %w", dm.storage, makeDirErr)
+		}
+		klog.InfoS("Backup storage directory created", "path", dm.storage)
+	} else {
+		klog.InfoS("Backup storage directory already existed", "path", dm.storage)
+	}
+
+	dest := dm.GetBackupPath(name)
+	tmp := dest + ".tmp"
+	old := dest + ".old"
+
+	// Make sure /storage/backup.tmp does not exist, so data isn't copied into that directory
+	if err := os.RemoveAll(tmp); err != nil {
+		return fmt.Errorf("failed to remove %s: %w", tmp, err)
+	}
+
+	if err := copyDataDir(tmp); err != nil {
+		return err
+	}
+
+	backupExists, err := dm.BackupExists(name)
+	if err != nil {
+		return err
+	} else if backupExists {
+		if err := renamePath(dest, old); err != nil {
+			return err
+		}
+		klog.InfoS("Temporarily renamed existing backup", "backup", dest, "renamed", old)
+	}
+
+	if err := renamePath(tmp, dest); err != nil {
+		klog.Errorf("Renaming path failed - renaming %s back and deleting %s: %v", old, tmp, err)
```

Comment:
It's going to be interesting to set up a test for this error case.

Reply:
We talked on slack about a way to generate unique names to avoid having to rename anything. Let's land this version and make that change in another PR.